rllib/utils/exploration/slate_epsilon_greedy.py

```diff
@@ -12,6 +12,64 @@
 
 
 class SlateEpsilonGreedy(EpsilonGreedy):
+    @override(EpsilonGreedy)
+    def _get_tf_exploration_action_op(
+        self,
+        action_distribution: ActionDistribution,
+        explore: Union[bool, TensorType],
+        timestep: Union[int, TensorType],
+    ) -> "tf.Tensor":
+
+        per_slate_q_values = action_distribution.inputs
+        all_slates = self.model.slates
+
+        exploit_indices = action_distribution.deterministic_sample()
+        exploit_action = tf.gather(all_slates, exploit_indices)
+
+        batch_size = tf.shape(per_slate_q_values)[0]
+        action_logp = tf.zeros(batch_size, dtype=tf.float32)
+
+        # Get the current epsilon.
+        epsilon = self.epsilon_schedule(
+            timestep if timestep is not None else self.last_timestep
+        )
+        # Mask out actions, whose Q-values are -inf, so that we don't
+        # even consider them for exploration.
+        random_valid_action_logits = tf.where(
+            tf.equal(per_slate_q_values, tf.float32.min),
+            tf.ones_like(per_slate_q_values) * tf.float32.min,
+            tf.ones_like(per_slate_q_values),
+        )
+        # A random action.
+        random_indices = tf.squeeze(
+            tf.random.categorical(random_valid_action_logits, 1), axis=1
+        )
+        random_actions = tf.gather(all_slates, random_indices)
+
+        chose_random = (
+            tf.random.uniform(
+                tf.stack([batch_size]), minval=0, maxval=1, dtype=tf.float32
+            )
+            < epsilon
+        )
+
+        # Pick either random or greedy.
+        action = tf.cond(
+            pred=tf.constant(explore, dtype=tf.bool)
+            if isinstance(explore, bool)
+            else explore,
+            true_fn=(lambda: tf.where(chose_random, random_actions, exploit_action)),
+            false_fn=lambda: exploit_action,
+        )
+
+        if self.framework in ["tf2", "tfe"] and not self.policy_config["eager_tracing"]:
+            self.last_timestep = timestep
+            return action, action_logp
+        else:
+            assign_op = tf1.assign(self.last_timestep, tf.cast(timestep, tf.int64))
```

Comment:
sorry, why do we need to keep track of and update timestep here?
will this timestep get restored if job crashes etc?

Reply:
Yes, it should. It's part of the exploration component's state.